watch_service.go

```diff
@@ -220,19 +222,29 @@ func (w *Watcher) Close() {
 }
 
 func (w *Watcher) addListenerChan(ch chan *WatchResult) {
-	// using `_` to prevent `nil` casting panic
-	chans, _ := w.updateListenerChans.Load().([]chan *WatchResult)
+	for {
+		// try to acquire write lock
+		if atomic.CompareAndSwapInt32(&w.listenerChansLock, 0, 1) {
+			// using `_` to prevent `nil` casting panic
+			chans, _ := w.updateListenerChans.Load().([]chan *WatchResult)
+
+			// get number of chans
+			n := len(chans)
 
-	// get number of chans
-	n := len(chans)
+			// copy-on-write
+			cow := make([]chan *WatchResult, n+1)
+			copy(cow, chans) // work even if chans == nil
+			cow[n] = ch
 
-	// copy-on-write
-	cow := make([]chan *WatchResult, n+1)
-	copy(cow, chans) // work even if chans == nil
-	cow[n] = ch
+			// store back
+			w.updateListenerChans.Store(cow)
 
-	// store back
-	w.updateListenerChans.Store(cow)
+			// reset lock
+			atomic.CompareAndSwapInt32(&w.listenerChansLock, 1, 0)
```

Comment:
nit: Could just do `w.listenerChansLock = 0`?

Reply:
perhaps `atomic.StoreInt32(&w.listenerChansLock, 0)`?

Reply:
yes, it could be but cause data race. Although this race is cause no bad effect, we should avoid it as user could run test with `-race` and fail!

Reply:
And `CAS` this case it better than Store. Store is spin CAS which cause more overhead.

Reply:
Ah didn't know about the internal implementation. Just thought that we don't have to use CAS if possible because it's just assigning the `0` value back to the memory.

Reply:
Anyway, thanks for the explanation.!

Reply:
Yeah, we dont have `volatile` type in go 😢 In view of racing test in go, when another routine try to `read` `listenerChansLock`. 
At the same time we write `listenerChansLock = 0`, even if there is no racing in write, it still considered race (read/write at the same time) and cause `test` fail :(

Reply:
Yeah, there's still a lot of stuff to learn. 😉

Reply:
Me too 😭